File: credit/credit.c
#include<cs50.h>
#include<stdio.h>

int numberOfDigitCheck(long long int n);

int luhnsAlgorithm(long long int n);

void checkCards(long long int n, int m);

int main(void) {
    long long int Credit_number;
    int m;
    Credit_number = get_long("Enter your credit number: ");
    m = numberOfDigitCheck(Credit_number);
    if (luhnsAlgorithm(Credit_number) % 10 == 0) {
        checkCards(Credit_number,m);
        printf("\n");
    }
    else {
        printf("INVALID");
        printf("\n");
    }
}

int numberOfDigitCheck(long long int n){
    int count=0;
    if (n == 0) {
        count = 1;
    }
    else {
        while (n != 0) {
            count++;
            n /= 10;
        }
    }
    return count;
}

int luhnsAlgorithm(long long int n){
    int sum_last = 0;
    int sum_second = 0;
    int digit_position = 1;
    while (n>0){
        int digit =n % 10;
        n/=10;
        if (digit_position % 2 == 0) {
            long long int double_value = digit * 2;
            long long int result = double_value % 10;
            sum_second += result + double_value/10;
        }
        else {
            sum_last += digit;
        }
        digit_position++;
    }
    int sum = sum_last + sum_second;
    return sum;
}

void checkCards(long long int n, int m){
    while (n >= 100) {
        n /= 10;
    }
    if(n==37&& m==15) {
        printf("AMEX");
    }
    else if(n==30&& m==14){
        printf("Diners Club");
    }
    else if(n==60&& m==16){
        printf("Discover");
    }
    else if(n==35&& m==16){
        printf("JCB");
    }
    else if((n==22||n==51||n==52)&& m==16){
        printf("MASTERCARD");
    }
    else if((n==22|| n== 55 || n == 51)&& m==16){
        printf("MASTERCARD");
    }
    else if((n==41|| n== 40 || n == 42 || n==49)&&(m==16||m==13)){
        printf("VISA");
    }
    else {
        printf("INVALID");
    }
}

File: python/sentimental-credit/credit.py
def main():
    while True:
        Num = input("Number: ")
        if len(Num) >= 13 and len(Num) <= 16:
            try:
                carNumber=int(Num)
                break
            except ValueError:
                break
        else:
            print('INVALID')
            break
    if luhnAlgorithm(Num):
        checkCards(Num)
    else:
        print('INVALID')

def checkCards(Num):
    m = len(str(Num))
    n = int(str(Num)[:2])
    if n == 37 and m == 15:
        print("AMEX")
    elif n == 30 and m == 14:
        print("Diners Club")
    elif n == 60 and m == 16:
        print("Discover")
    elif n == 35 and m == 16:
        print("JCB")
    elif (n == 22 or n == 51 or n == 52) and m == 16:
        print("MASTERCARD")
    elif (n == 22 or n == 55 or n == 51) and m == 16:
        print("MASTERCARD")
    elif (n == 41 or n == 40 or n == 42 or n == 49) and (m == 16 or m == 13):
        print("VISA")
    else:
        print("INVALID")

def luhnAlgorithm(Num):
    second_to_last=[]
    others=[]
    for i, digit_char in enumerate(Num):
        digit = int(digit_char)
        #check if the distance from i to end of string is odd or even
        #if it's odd -> mean it's not the second to last digit
        if(len(Num)-i+1)%2==0:
            others.append(digit)
        else:
            second = digit*2
            if second>= 10:
                second_to_last.extend([int(digits) for digits in str(second)])
            else:
                second_to_last.append(second)
    if (sum(second_to_last) + sum(others)) % 10 == 0:
        return True
    else:
        return False
main()
